File: sample/CSharp/SimpleRankingsServer.cs
using System;
using System.Text;
using System.IO;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Json;

namespace SimpleRankingsServer
{
    [DataContract]
    public class Data<T>
    {
        [DataMember]
        private string userId { get; set; }

        [DataMember]
        private string utcDate { get; set; }


        [DataMember(Name = "id")]
        public long Id { get; private set; }


        [DataMember(Name = "values")]
        public T Values { get; private set; }

        [IgnoreDataMember]
        public Guid UserId { get; private set; }

        [IgnoreDataMember]
        public DateTime UTCDate { get; private set; }

        [OnDeserialized]
        private void OnDeserialized(StreamingContext s)
        {
            UserId = Guid.Parse(userId);
            UTCDate = DateTime.ParseExact(utcDate, "yyyy/MM/dd HH:mm:ss", null);
        }

        public override string ToString()
        {
            return $"{{ \"id\" : {Id}, \"userId\" : {userId}, \"utcDate\" : {utcDate}, \"values\" : {Values} }}";
        }
    }

    public static class JsonUtils
    {
        public static string Serialize<T>(T obj)
        {
            using (var ms = new MemoryStream())
            {
                var serializer = new DataContractJsonSerializer(typeof(T));
                serializer.WriteObject(ms, obj);
                return Encoding.UTF8.GetString(ms.ToArray());
            }
        }

        public static T Deserialize<T>(string json)
        {
            using (var ms = new MemoryStream(Encoding.UTF8.GetBytes(json)))
            {
                var serializer = new DataContractJsonSerializer(typeof(T));
                return (T)serializer.ReadObject(ms);
            }
        }
    }

    public class Client : IDisposable
    {
        private readonly string url;
        private readonly HttpClient client = new HttpClient();

        public Client(string url, string username, string password)
        {
            this.url = url;
            var parameter = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{username}:{password}"));
            client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Basic", parameter);
        }

        public void Dispose()
        {
            client.Dispose();
        }

        [DataContract]
        private class InsertParam<T>
        {
            [DataMember(Name = "table")]
            public string Table { get; set; }

            [DataMember(Name = "userId")]
            public string UserIdStr { get; set; }

            [DataMember(Name = "values")]
            public T Values { get; set; }

            [IgnoreDataMember]
            private Guid userId;

            [IgnoreDataMember]
            public Guid UserId
            {
                get => userId;
                set
                {
                    userId = value;
                    UserIdStr = value.ToString();
                }
            }
        }

        [DataContract]
        private class InsertResult
        {
            [DataMember(Name = "id")]
            public long Id { get; set; }
        }

        public async Task<long> InsertAsync<T>(string tableName, Guid userId, T data)
        {
            var json = JsonUtils.Serialize(new InsertParam<T> { Table = tableName, UserId = userId, Values = data });
            using (var content = new StringContent(json, Encoding.UTF8, @"application/json"))
            {
                var result = await client.PostAsync(url, content);
                var resString = await result.Content.ReadAsStringAsync();

                if (result.IsSuccessStatusCode)
                {
                    return JsonUtils.Deserialize<InsertResult>(resString).Id;
                }

                throw new Exception($"{result.StatusCode}:{resString}");
            }
        }

        public async Task<IReadOnlyList<Data<T>>> SelectAsync<T>(string table, string orderBy = null, bool isDescending = true, int limit = 100)
        {
            var param = new Dictionary<string, string>();
            param.Add("table", table);
            param.Add("isDescending", isDescending.ToString());
            param.Add("limit", limit.ToString());
            if (orderBy != null) param.Add("orderBy", orderBy);

            var result = await client.GetAsync($"{url}?{await new FormUrlEncodedContent(param).ReadAsStringAsync()}");
            var resString = await result.Content.ReadAsStringAsync();

            if (result.IsSuccessStatusCode)
            {
                return JsonUtils.Deserialize<Data<T>[]>(resString);
            }

            throw new Exception($"{result.StatusCode}:{resString}");
        }
    }
}

File: sample/FSharp/SimpleRankingsServer.fs
module SimpleRankingsServer

open System
open System.Text
open System.Collections.Generic
open System.Net.Http
open System.Net.Http.Headers
open FSharp.Json
open System.ComponentModel

type DateTimeString() =
  interface ITypeTransform with
    member __.targetType() = (fun _ -> typeof<string>) ()
    member __.toTargetType value = (fun (v: obj) ->
      (v :?> DateTime).ToString "yyyy/MM/dd HH:mm:ss" :> obj) value
    
    member __.fromTargetType value = (fun (v: obj) ->
      DateTime.ParseExact(v:?> string, "yyyy/MM/dd HH:mm:ss", null) :> obj) value

type Data<'a> = {
  id : int64
  userId : Guid
  values : 'a

  [<JsonField(Transform=typeof<DateTimeString>)>]
  utcDate : DateTime
}

module Internal =
  type InsertParam<'a> = {
    table : string
    userId : Guid
    values : 'a
  }

  type InsertResult = { id : int64 }

open Internal

let private jsonConfig = JsonConfig.create(allowUntyped = true)

type Client(url : string, usrename, password) =
  let client = new HttpClient()

  do
    let parameter = Convert.ToBase64String(Encoding.UTF8.GetBytes(sprintf "%s:%s" usrename password));
    client.DefaultRequestHeaders.Authorization <- AuthenticationHeaderValue("Basic", parameter);

  interface IDisposable with
    member __.Dispose() = client.Dispose()

  member __.AsyncInsert (tableName, userId, data) : Async<int64> =
    async {
      let json = Json.serializeEx jsonConfig { table = tableName; userId = userId; values = data }
      use content = new StringContent(json, Encoding.UTF8, @"application/json")

      let! result = client.PostAsync(url, content) |> Async.AwaitTask
      let! resString = result.Content.ReadAsStringAsync() |> Async.AwaitTask

      if result.IsSuccessStatusCode then
        return Json.deserialize<InsertResult>(resString).id
      else
        return failwithf "%A:%s" result.StatusCode resString
    }

  member __.AsyncSelect (tableName: string, ?orderBy : string, ?isDescending : bool, ?limit : int) : Async<Data<'a>[]> =
    async {
      let param = Dictionary<string, string>()
      let inline add s x =
        x |> Option.iter(fun y -> param.Add(s, y.ToString()))

      param.Add("table", tableName)
      orderBy |> add "orderBy"
      isDescending |> add "isDescending"
      limit |> add "limit"

      let! paramStr = (new FormUrlEncodedContent(param)).ReadAsStringAsync() |> Async.AwaitTask
      let! result = client.GetAsync(sprintf "%s?%s" url paramStr) |> Async.AwaitTask
      let! resString = result.Content.ReadAsStringAsync() |> Async.AwaitTask

      if result.IsSuccessStatusCode then
        return Json.deserializeEx<Data<'a>[]> jsonConfig resString
      else
        return failwithf "%A:%s" result.StatusCode resString
    }
